src/lib.rs

```diff
@@ -1452,6 +1452,141 @@ fn collect_runs(
+        Ok(nodes) => nodes,
+        Err(_err) => {
+            return Err(DAGHasCycle::new_err("Sort encountered a cycle"))
+        }
+    };
+
+    // Utility for ensuring pending_list has the color index
+    macro_rules! ensure_vector_has_index {
+        ($pending_list: expr, $block_id: expr, $color: expr) => {
+            if $color >= $pending_list.len() {
+                $pending_list.resize($color + 1, Vec::new());
+                $block_id.resize($color + 1, None);
+            }
+        };
+    }
+
+    for node in nodes {
+        if special_filter_node(&graph.graph[node])? {
+            let raw_edges = graph
+                .graph
+                .edges_directed(node, petgraph::Direction::Outgoing);
+
+            // Filter all edges that yield non-negative colors from color_fn
+            let colors = raw_edges
+                .filter_map(|edge| {
+                    let edge_weight = edge.weight();
+                    let color_result = color_edge(edge_weight);
+                    if color_result.is_ok() {
+                        let color = color_result.unwrap_or_default();
+                        if color >= 0 {
+                            Some(Ok(color))
+                        } else {
+                            None
+                        }
+                    } else {
+                        Some(color_result)
+                    }
+                })
+                .collect::<PyResult<Vec<i32>>>()?;
+
+            if colors.len() <= 2 && filter_node(&graph.graph[node])? {
+                if colors.len() == 1 {
+                    let c0 = colors[0] as usize;
+                    ensure_vector_has_index!(pending_list, block_id, c0);
+                    if let Some(c0_block_id) = block_id[c0] {
+                        block_list[c0_block_id]
+                            .push(graph.graph[node].clone_ref(py));
+                    } else {
+                        pending_list[c0].push(graph.graph[node].clone_ref(py));
+                    }
+                } else {
+                    let c0 = colors[0] as usize;
+                    let c1 = colors[1] as usize;
+                    ensure_vector_has_index!(pending_list, block_id, c0);
+                    ensure_vector_has_index!(pending_list, block_id, c1);
+
+                    if block_id[c0].is_some()
+                        && block_id[c1].is_some()
+                        && block_id[c0].unwrap_or_default()
+                            == block_id[c1].unwrap_or_default()
+                    {
+                        block_list[block_id[c0].unwrap_or_default()]
+                            .push(graph.graph[node].clone_ref(py));
+                    } else {
+                        let mut new_block: Vec<PyObject> = Vec::new();
+                        if !pending_list[c0].is_empty() {
+                            new_block.extend(pending_list[c0].iter().cloned());
+                            pending_list[c0].clear();
```

Comment:
I think this is equivalent to 
```suggestion
                            new_block.append(&mut pending_list[c0]);
```
and we avoid a clone (it shouldn't be expensive but nevertheless)